### PR TITLE
Updates for the next Doctum version

### PR DESCRIPTION
**Pull request by williamdes**
Fixes: https://github.com/silverstripe/api.silverstripe.org/issues/59
Closes: https://github.com/silverstripe/api.silverstripe.org/pull/93

### CODE REVIEW

File: src/RemoteRepository/SilverStripeRemoteRepository.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        // $this->localPath is the root path to packages
+        // Example: /mnt/Dev/@code-lts/@doctum-fork/api.silverstripe.org/data/packages/
+        // $this->name is empty, and not used
+        // $relativePath for the current file
+        // Example "silverstripe/admin/code/AdminRootController.php"
+        $pathParts = explode('/', $relativePath, 3);
+        // Example: [silverstripe, admin, code/AdminRootController.php]
+        $this->name = $pathParts[0] . '/' . $pathParts[1];
+        $url = 'https://github.com/' . $this->name . '/blob/' . $this->buildProjectPath($projectVersion, $relativePath);
```
Comment (andrewandante) [inline src/RemoteRepository/SilverStripeRemoteRepository.php:20]:
```suggestion
        $url = 'https://github.com/' . $this->name . '/blob/' . $this->buildProjectPath($projectVersion, $pathParts[2]);
```

@williamdes just checked this out locally, the generated links aren't quite right, I'm getting:

`https://github.com/silverstripe/framework/blob/4/silverstripe/framework/src/Core/Config/Configurable.php#L20` for example  - note that `silverstripe/framework` is in there twice. Think the above would fix it?

Comment (andrewandante) [reply src/RemoteRepository/SilverStripeRemoteRepository.php:20]: Hmm no well that fixes one part - the duplication - but in this case, the repo URL for silverstripe/admin is `silverstripe/silverstripe-admin/` - so using the checked-out path isn't going to work there. I'll see if I can work something out.

Comment (andrewandante) [reply src/RemoteRepository/SilverStripeRemoteRepository.php:20]: Alright, this seems to work:

```php
        $packageConfig = Config::getConfig()['packages'][$this->name];
        $rootPath = $packageConfig['repository'];
        if (substr($rootPath, -4) === '.git') {
            $rootPath = substr($rootPath, 0, -4);
        }
        if (isset($packageConfig['versionmap'])) {
            $versionMaps = Config::getConfig()['versionmaps'];
            $projectVersion = $versionMaps[$packageConfig['versionmap']][(string) $projectVersion];
        }
        $url = $rootPath . '/blob/' . $this->buildProjectPath($projectVersion, $pathParts[2]);
```

Takes into account both the correct URL for the repo, the correct version to use to align everything, and removing the stray `.git` extension on the repo

**Comment by williamdes** [comment]
Thank you @andrewandante !
I added a unit test to prevent regressions and be sure the code works for ever ^^
And I added you as a co-author

**Comment by andrewandante** [comment]
Test is an excellent addition, thanks!

Had an issue building locally, I've fixed it by adding `"3": null` to the `versionmap` for `starts-at-three` - graphql was throwing me an error otherwise.

**Comment by williamdes** [comment]
> Test is an excellent addition, thanks!
> 
> Had an issue building locally, I've fixed it by adding `"3": null` to the `versionmap` for `starts-at-three` - graphql was throwing me an error otherwise.

Ha, is it a hack or a bug fix?
Could you send me the patch?

**Comment by andrewandante** [comment]
I think it's a bugfix - there was no graphql module associated with SS3, but it's trying to figure one out. This is the change to `conf/doctum.json`:

```diff
"versionmaps": {
        "starts-at-one": {
            "master": "master",
            "4": "1",
            "3": null
        },
        "starts-at-three": {
        	"master": "master",
+        	"4": "3",
+           "3": null
        }
    }
```

You can see it matches the pattern from the `starts-at-one` block so I think it makes sense

**Comment by williamdes** [comment]
Good catch, I added 2 test cases and used the `??` operator for a clean fix just in case. All covered by tests of course ;p

Closing back my workstation to try to sleep :laughing:

**Comment by andrewandante** [comment]
Looks good to me and works locally 👍

**Comment by chillu** [comment]
Phew, that's a lot of changes required for a minor release! But I guess we're "power users" in this aspect (file path mapping). Thanks for keeping on improving this Willian, and happy 2021 to you! I haven't gone through the details on this PR, but if Andrew approves I'm happy with it as well.

**Comment by williamdes** [comment]
> Phew, that's a lot of changes required for a minor release! But I guess we're "power users" in this aspect (file path mapping). Thanks for keeping on improving this Willian, and happy 2021 to you! I haven't gone through the details on this PR, but if Andrew approves I'm happy with it as well.

Ha I made a lot of changes for you upstream also you know ;)
This one can be merged: https://github.com/silverstripe/api.silverstripe.org/pull/95 and will have effect after this one also is
Just pushing another commit to upgrade the Doctum version

**Review by andrewandante — APPROVED** [review]

**Comment by williamdes** [comment]
Rebased and just added 822c2b33551611a950057ad903f1c4c22fe42d90

**Comment by andrewandante** [comment]
Amazing, I was just doing that myself - I've been granted write access so if you are now happy, I'm happy too 👍

**Comment by andrewandante** [comment]
Amazing work, thank you so much!

**Pull request merged**